File: C#/E3.AddFunctionality/S4.Refactor/Calculator.cs
using System;

namespace TDDLegacyCode.E3.AddFunctionality.S4.Refactor
{
    public class Calculator
    {
        string display = "";
        int lastArgument = 0;
        int result = 0;
        bool newArgument = false;
        bool shouldReset = true;

        OperationType lastOperation;


        public void Press(string key)
        {
            if (key.Equals("C"))
            {
                display = "";
                return;
            }
            switch (key)
            {
                case "+":
                    HandleOpKey(OperationType.Plus);
                    break;
                case "-":
                    HandleOpKey(OperationType.Minus);
                    break;
                case "/":
                    HandleOpKey(OperationType.Div);
                    break;
                case "=":
                    HandleEquals(); 
                    break;
                default:
                    HandleNumbers(key); 
                    break;
            }
        }

        private void HandleNumbers(string key)
        {
            if (shouldReset)
            {
                display = "";
                shouldReset = false;
            }
            if (newArgument)
            {
                display = "";
                newArgument = false;
            }
            display += key;
        }

        private void HandleEquals()
        {
            int currentArgument = ParseKeyNumber();
            if (lastOperation == OperationType.Plus)
            {
                display = (lastArgument + currentArgument).ToString();
            }

            if (lastOperation == OperationType.Div && currentArgument == 0)
            {
                display = "Division By Zero Error";
            }
            shouldReset = true;
        }

        private void HandleOpKey(OperationType op)
        {
            lastOperation = op;
            lastArgument = ParseKeyNumber();
            newArgument = true;
        }

        private int ParseKeyNumber()
        {
            if (display.Equals(""))
                return 0;
            return Int32.Parse(display);
        }

        public String GetDisplay()
        {
            if (display.Equals(""))
                return "0";
            return display;
        }
    }

}

File: Python/e3.addfunctionality/s4.refactor/calculator.py
from enum import Enum


class OperationType(Enum):
    Plus = 1
    Div = 2
    Minus = 3


class Calculator:
    display = ""
    lastArgument = 0
    result = 0
    newArgument = False
    shouldReset = True
    lastOperation = OperationType.Plus

    def press(self, key):
        if key == "C":
            self.display = ""
            return

        match key:
            case "+":
                self.handleOpKey(OperationType.Plus)
            case "/":
                self.handleOpKey(OperationType.Div)
            case "-":
                self.handleOpKey(OperationType.Minus)
            case "=":
                self.handle_equals()
            case _:
                self.handle_number(key)

    def handle_number(self, key):
        if self.shouldReset:
            self.display = ""
            self.shouldReset = False
        if self.newArgument:
            self.display = ""
            self.newArgument = False
        self.display += key

    def handle_equals(self):
        currentArgument = self.parseKeyNumber()
        if self.lastOperation == OperationType.Plus:
            self.display = str(self.lastArgument + currentArgument)
        if self.lastOperation == OperationType.Div and currentArgument == 0:
            self.display = "Division By Zero Error"
        self.shouldReset = True

    def get_display(self):
        if self.display == "":
            return "0"
        return self.display

    def parseKeyNumber(self):
        if self.display == "":
            return 0
        return int(self.display)

    def handleOpKey(self, operationType):
        self.lastOperation = operationType
        self.lastArgument = self.parseKeyNumber()
        self.newArgument = True
